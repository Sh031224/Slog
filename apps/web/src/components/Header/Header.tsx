import * as S from "./index.style";

const Header: React.FC = () => {
  return <S.Wrapper>Header</S.Wrapper>;
};

export default Header;
