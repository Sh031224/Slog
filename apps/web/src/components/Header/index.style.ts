import styled from "styled-components";

export const Wrapper = styled.header`
  width: 100px;
`;
