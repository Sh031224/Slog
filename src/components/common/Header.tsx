import React from "react";
import HeadRoom from "react-headroom";
import styled from "styled-components";
import logo from "assets/images/logo.png";
import useHeader from "hooks/useHeader";
import ScrollToTop from "react-scroll-to-top";
import { GrLinkTop } from "react-icons/gr";
import { ReactCompnent as SearchImg } from "../../assets/images/search.svg";
import useFacebookLogin from "hooks/useFacebookLogin";
import FacebookLogin from "react-facebook-login/dist/facebook-login-render-props";
import Link from "next/link";

const Header: React.FC = () => {
  const {
    search,
    isToggle,
    searchEl,
    changeIsToggle,
    onChangeSearch,
    onSubmitSearch,
    onKeypressSearch
  } = useHeader();

  const { login, appId, tryLogin, tryLogout } = useFacebookLogin();

  return (
    <>
      <HeadRoom>
        <HeaderWrapper>
          <HeaderContainer>
            <HeaderNav>
              <Link href="/">
                <a>
                  <HeaderLogo src={logo} alt={"Slog"} />
                </a>
              </Link>
              <HeaderUtils>
                <HeaderSearch onClick={changeIsToggle} isToogle={isToggle}>
                  <HeaderSearchInput
                    ref={searchEl}
                    isToogle={isToggle}
                    type="text"
                    value={search}
                    onChange={onChangeSearch}
                    onKeyPress={onKeypressSearch}
                    placeholder={"검색어를 입력해주세요."}
                  />
                  <SearchImg onClick={onSubmitSearch} />
                </HeaderSearch>
                {login ? (
                  <HeaderLink onClick={tryLogout}>로그아웃</HeaderLink>
                ) : (
                  <FacebookLogin
                    appId={appId}
                    callback={tryLogin}
                    fields="name,email"
                    disableMobileRedirect={true}
                    render={(renderProps: any) => {
                      return <HeaderLink onClick={renderProps.onClick}>로그인</HeaderLink>;
                    }}
                  />
                )}
              </HeaderUtils>
            </HeaderNav>
          </HeaderContainer>
        </HeaderWrapper>
      </HeadRoom>
      <ScrollToTop smooth component={<GrLinkTop />} />
    </>
  );
};

const HeaderWrapper = styled.header`
  top: 0;
  width: 100%;
  display: flex;
  justify-content: center;
  padding: 0.5rem;
  box-shadow: ${({ theme }) => theme.colors.sdLightBlack} 0px 2px 10px 0px;
  background-color: ${({ theme }) => theme.colors.bgWhite};
`;

const HeaderContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  max-width: 1200px;
  padding: 0 1rem;
  ${({ theme }) => theme.device.mobile} {
    padding: 0 0.5rem;
  }
`;

const HeaderNav = styled.nav`
  display: flex;
  justify-content: space-between;
  align-items: center;
  a {
    display: flex;
    align-items: center;
  }
`;

const HeaderLogo = styled.img`
  cursor: pointer;
  width: 2.5rem;
  ${({ theme }) => theme.device.mobile} {
    width: 2rem;
  }
`;

const HeaderUtils = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;

const HeaderLink = styled.span`
  color: ${({ theme }) => theme.colors.ftGray};
  font-weight: 600;
  font-size: 0.95rem;
  cursor: pointer;
  ${({ theme }) => theme.device.mobile} {
    font-size: 0.8rem;
  }
`;

const HeaderSearch = styled.div<{ isToogle: boolean }>`
  position: relative;
  width: 1.8rem;
  height: 1.8rem;
  overflow: hidden;
  border-radius: 1.8rem;
  margin-right: 1rem;
  max-width: 195px;
  border: ${({ theme }) => theme.colors.bdLightGray} 1px solid;
  transition: all ease-in-out 0.5s 0s;
  display: flex;
  align-items: center;
  ${({ isToogle }) =>
    isToogle
      ? "justify-content: flex-start; width: 11.5rem;"
      : "justify-content: center; width: 1.8rem;"}
  ${({ theme }) => theme.device.mobile} {
    width: ${({ isToogle }) => (isToogle ? "10rem" : "1.5rem")};
    height: 1.5rem;
    border-radius: 1.5rem;
  }

  & > svg {
    width: 0.8rem;
    height: 0.8rem;
    cursor: pointer;
    opacity: 0.5;
  }
`;

const HeaderSearchInput = styled.input<{ isToogle: boolean }>`
  color: ${({ theme }) => theme.colors.ftGray};
  display: ${({ isToogle }) => (isToogle ? "inline-block" : "none")};
  width: calc(100% - 1.5rem);
  max-width: 165px;
  height: 1.8rem;
  padding-left: 1rem;
  padding-right: 0.5rem;
  border: none;
  outline: none;
  font-size: 0.8rem;
  font-weight: 500;
  ${({ theme }) => theme.device.mobile} {
    height: 1.5rem;
    width: calc(100% - 2rem);
    font-size: 0.7rem;
  }
`;

export default React.memo(Header);
