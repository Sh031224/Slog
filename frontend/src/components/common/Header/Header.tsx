import React, { Dispatch, MutableRefObject } from "react";
import HeadRoom from "react-headroom";
import "./Header.scss";
import logo from "../../../assets/images/logo.png";
import ScrollToTop from "react-scroll-to-top";
import FacebookLogin from "react-facebook-login/dist/facebook-login-render-props";
import { appId } from "../../../config/index.json";
import { GrLinkTop } from "react-icons/gr";
import {
  ReactFacebookFailureResponse,
  ReactFacebookLoginInfo
} from "react-facebook-login";
import HeaderSearch from "./HeaderSearch";

interface HeaderProps {
  searchEl: MutableRefObject<any>;
  inputEl: MutableRefObject<any>;
  search: string;
  setSearch: Dispatch<React.SetStateAction<string>>;
  login: boolean;
  tryLogin: (res: FacebookLoginInfo | FacebookFailureResponse) => void;
  tryLogout: () => void;
  searchSubmit: () => void;
  clickLogo: () => void;
}

interface FacebookLoginInfo extends ReactFacebookLoginInfo {
  accessToken: string;
}

interface FacebookFailureResponse extends ReactFacebookFailureResponse {
  accessToken?: string;
}

const Header = ({
  searchEl,
  inputEl,
  search,
  setSearch,
  login,
  tryLogout,
  tryLogin,
  searchSubmit,
  clickLogo
}: HeaderProps) => {
  return (
    <>
      <HeadRoom>
        <header className="header">
          <div className="header-container">
            <div className="header-container-main">
                <img
                  className="header-container-main-logo"
                  src={logo}
                  alt="logo"
                  onClick={clickLogo}
                />
              <div className="header-container-main-util">
                <HeaderSearch
                  search={search}
                  setSearch={setSearch}
                  searchEl={searchEl}
                  inputEl={inputEl}
                  searchSubmit={searchSubmit}
                />
                {login ? (
                  <span
                    onClick={tryLogout}
                    className="header-container-main-util-login"
                  >
                    로그아웃
                  </span>
                ) : (
                  <FacebookLogin
                    appId={appId}
                    fields="name,email"
                    disableMobileRedirect={true}
                    callback={tryLogin}
                    render={(renderProps: any) => {
                      return (
                        <div
                          className="login-container-button"
                          onClick={renderProps.onClick}
                        >
                          <span className="header-container-main-util-login">
                            로그인
                          </span>
                        </div>
                      );
                    }}
                  />
                )}
              </div>
            </div>
          </div>
        </header>
      </HeadRoom>
      <ScrollToTop smooth component={<GrLinkTop />} />
    </>
  );
};

export default Header;
