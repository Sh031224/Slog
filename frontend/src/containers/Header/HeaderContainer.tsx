import { observer } from "mobx-react";
import React, { useCallback, useEffect, useRef, useState } from "react";
import axios from "axios";
import { NotificationManager } from "react-notifications";
import { useRouter } from "next/router";
import cookies from "js-cookie";
import firebase from "firebase/app";
import option from "../../config/firebase.json";
import "firebase/messaging";
import useStore from "lib/hooks/useStore";
import { FacebookFailureResponse, FacebookLoginInfo } from "types/FacebookType";
import Header from "components/common/Header";

interface HeaderContainerProps {
  token?: string;
}

const HeaderContainer = ({ token }: HeaderContainerProps) => {
  const { store } = useStore();
  const {
    handleUser,
    haldleAdminFalse,
    login,
    handleLoginChange,
    handleLogin,
    handleFcm
  } = store.UserStore;
  const { handleIsClickedLogo } = store.HistoryStore;
  const searchEl = useRef<HTMLElement>(null);
  const inputEl = useRef<HTMLElement>(null);

  const router = useRouter();

  const [search, setSearch] = useState("");

  useEffect(() => {
    if (token) {
      axios.defaults.headers.common["access_token"] = token;
    } else {
      axios.defaults.headers.common["access_token"] = cookies.get(
        "access_token"
      );
    }
  }, [token]);

  const tryLogin = async (res: FacebookLoginInfo | FacebookFailureResponse) => {
    await handleLogin(res.accessToken!)
      .then(async (response: any) => {
        const today = new Date();
        today.setDate(today.getDate() + 30);
        cookies.set("access_token", response.data.access_token, {
          expires: today,
          path: "/"
        });
        axios.defaults.headers.common["access_token"] =
          response.data.access_token;
        NotificationManager.success("로그인 되었습니다.", "Success");
        handleAll(response.data.access_token);
      })
      .catch(() => {
        haldleAdminFalse();
        NotificationManager.error("로그인에 실패하였습니다.", "Error");
      });
  };

  const tryLogout = useCallback(() => {
    cookies.remove("access_token");
    handleLoginChange(false);
    axios.defaults.headers.common["access_token"] = "";
    haldleAdminFalse();
  }, []);

  const searchSubmit = useCallback(() => {
    if (searchEl.current) {
      if (
        searchEl.current.classList.contains(
          "header-container-main-util-search-active"
        )
      ) {
        if (search !== "") {
          router.push(`/?search=${search}`);
        } else {
          router.push("/");
        }
      }
    }
  }, [search]);

  const getFcmToken = useCallback(async () => {
    if (!firebase.apps.length) {
      firebase.initializeApp(option);
    }
    const token = await firebase.messaging().getToken();
    handleFcm(token);
  }, [handleFcm]);

  const requestNotification = useCallback(() => {
    if ("Notification" in window) {
      try {
        Notification.requestPermission().then(
          (permission: NotificationPermission) => {
            if (permission === "granted") {
              getFcmToken();
            }
          }
        );
      } catch (error) {
        // Safari doesn't return a promise for requestPermissions and it
        // throws a TypeError. It takes a callback as the first argument
        // instead.
        if (error instanceof TypeError) {
          Notification.requestPermission(
            (permission: NotificationPermission) => {
              if (permission === "granted") {
                getFcmToken();
              }
            }
          );
        } else {
          throw error;
        }
      }
    }
  }, [getFcmToken]);

  const handleAll = useCallback(async (access_token: string) => {
    const token = axios.defaults.headers.common["access_token"];
    if (token) {
      await handleUser(access_token).catch((err) => {
        if (err.message === "401") {
          tryLogout();
        }
      });
      requestNotification();
    }
  }, []);

  const handleLoginCallback = useCallback(() => {
    const token = axios.defaults.headers.common["access_token"];

    if (!login) {
      if (token) {
        handleLoginChange(true);
        handleUser(token).catch((err) => {
          if (err.message === "Error: Request failed with status code 401") {
            tryLogout();
          } else {
            NotificationManager.error("오류가 발생하였습니다.", "Error");
          }
        });
        requestNotification();
      }
    }
  }, [login, requestNotification]);

  useEffect(() => {
    handleLoginCallback();
  }, [handleLoginCallback]);

  const clickLogo = () => {
    handleIsClickedLogo(true);
    router.push("/")
  }

  return (
    <>
      <Header
        search={search}
        setSearch={setSearch}
        searchEl={searchEl}
        inputEl={inputEl}
        login={login}
        tryLogin={tryLogin}
        tryLogout={tryLogout}
        searchSubmit={searchSubmit}
        clickLogo={clickLogo}
      />
    </>
  );
};

export default React.memo(observer(HeaderContainer));
