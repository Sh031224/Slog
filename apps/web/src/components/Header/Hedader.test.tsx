import { render } from "@testing-library/react";
import Header from "./Header";

describe("Header.tsx", () => {
  const setup = (props = {}) => {
    return render(<Header {...props} />);
  };

  it("matches snapshot", () => {
    const { container } = setup();

    expect(container.querySelector("header")?.textContent).toBe("Header");
  });
});
