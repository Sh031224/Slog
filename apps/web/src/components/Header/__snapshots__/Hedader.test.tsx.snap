// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header.tsx matches snapshot 1`] = `
<div>
  <header
    class="sc-bczRLJ cETFyA"
  >
    Header
  </header>
</div>
`;
